### ISSUE TITLE
HTTPClientRequest.writeFormBody switched to chunked encoding

### ISSUE THREAD

**Issue by punkUser** [rejectedsoftware/vibe.d#1779]
I'm not sure this is an "issue" per se, but it caused me some hard to track down regressions in one of my applications so it's likely worth documenting at the very least:

A recent-ish change to HTTPClientRequest.writeFormBody indirectly caused it to switch to "chunked" encoding for the request, due to querying the bodyWriter before setting a content length. Strangely, I can't find this code in the repo, but in my dub-acquired 0.8.0-beta.7 I see:
```
/** Writes the request body as form data.
*/
void writeFormBody(T)(T key_value_map)
{
	import vibe.inet.webform : formEncode;
	import vibe.stream.wrapper : streamOutputRange;

	headers["Content-Type"] = "application/x-www-form-urlencoded";
	auto dst = streamOutputRange(bodyWriter);
	() @trusted { return &dst; } ().formEncode(key_value_map);
}
```

Interestingly, form.d has a different version that likely doesn't have the same issue:
```
/**
	Writes a `vibe.http.client.HTTPClientRequest` body as URL encoded form data.
*/
void writeFormBody(HTTPClientRequest req, in string[string] form)
{
	import vibe.http.form;
	import vibe.stream.wrapper;

	StringLengthCountingRange len;
	writeFormData(&len, form);
	req.contentType = "application/x-www-form-urlencoded";
	req.contentLength = len.count;
	auto rng = streamOutputRange(req.bodyWriter);
	writeFormData(&rng, form);
}
```

Thus it might be that my repo is strangely out of date or something, but as I mentioned since this is something that didn't fail to compile upon upgrading my vibe version but broke the application (a number of servers my application talks to don't support chunked encoding at all) I figured it was something at least mentioning.

Thanks in advance!

**Comment by s-ludwig** [comment]
Unfortunately for some reason I didn't mention a context for the change, but the responsible commit is b10ae4750099a6f082b5d6a714a662d7bb8fd02c. The whole `vibe.http.form` module should be deprecated, though. The core functionality is nowadays in `vibe.inet.webform` and the convenience wrappers can go into `vibe.http.client` directly.

I'll fix this and put up another RC for 0.8.0.

**Comment by punkUser** [comment]
Ah thanks for the pointer to vibe.inet.webform, I'll take a look. Is there any example code that uses it offhand? Cheers!

**Comment by s-ludwig** [comment]
There is a small example in the docs: http://vibed.org/api/vibe.inet.webform/formEncode
Other than that, it's just used directly by the above code in `HTTPClientRequest.writeFormBody`. I've opened a PR and will tag a new RC for 0.8.0 once that is merged.

**Comment by s-ludwig** [comment]
Fixed by #1819.
